File: binary-search/binary-search.js
// https://leetcode.com/problems/binary-search/description/

class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number}
     */

    //start off with an arr of nums called numms
    //also have a target num value to search target
    //end with a num as a result (-1 if not found or index if found)
    //nums = [-1,0,2,4,6,8], target = 4 => 3
    //implement binary search to find target
    search(nums, target) {
        let L = 0, R = nums.length - 1;

        // let mid = Math.floor((L + R) / 2);

        while(L <= R){
            let mid = Math.floor((L + R) / 2);
            
            if(target > nums[mid]){
                L = mid + 1
            } else if (target < nums[mid]){
                R = mid - 1
            } else {
                return mid;
            }
        }
        return -1;
    }
}

File: python/binary-search/binary_search.py
# starting out with a list of nums and nums
# also have a num called target
# we want to find the index of the target and return index
# nums = [-1,0,2,4,6,8], target = 4 => 3
# adjust pointers based on the status between mid & target
# implement binary search return index of target or -1
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            mid = (left + right) // 2
            if nums[mid] < target:
                left = mid + 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                return mid
        return -1